File: tools/docs-guides.mjs
#!/usr/bin/env node
// docs-guides.mjs — FT4: guides keep-list (decision record), content-map
// generator, and the shared /docs/ route resolver used by BOTH the build
// (tools/docs-build.mjs rewrites/greys links) and the link checker
// (tools/docs-links.mjs) — one table, no drift between the two.
//
// Keep-list decision (2026-08-22, recorded in docs/content-map.json):
//   adapted : installation (upstream index is CLI/create-only → authored
//             vanilla rewrite docs/content/installation.mdx), dark-mode
//             (upstream index is framework cards → authored CSS-variables
//             rewrite docs/content/dark-mode.mdx), utils/shimmer +
//             utils/scroll-fade (mirrored; Installation section replaced —
//             the utilities ship precompiled inside dist/out.css)
//   mirror  : rtl (load-bearing: 56 radix pages link /docs/rtl), helpers/
//             ai-sdk + helpers/tanstack-ai (@shadcn/helpers reference; kept
//             per keep-list, previews are base-style → unavailable)
//   pruned  : forms / react / registry (React-only), changelog, (root), and
//             every framework sub-page (installation/*, dark-mode/*, rtl/*) —
//             links to pruned targets render as unavailable-link spans
//
// Guide ComponentPreview names are enumerated here with the same
// fence-stripped tag-scoped scan as docs-catalog.mjs (that tool stays
// radix-only by design; FT4 guide enumeration lives here, per PLAN FT4).
import { readFileSync, writeFileSync, mkdirSync, existsSync } from 'node:fs'
import { basename, join } from 'node:path'

const UP = '.upstream/shadcn-ui/apps/v4/content/docs'

export const GUIDES = [
  {
    slug: 'introduction',
    route: '/docs/introduction',
    title: 'Introduction',
    source: 'docs/content/introduction.mdx',
    disposition: 'adapted',
    notes: 'FT8: vanilla rewrite of upstream (root)/index.mdx; pure prose + 1 Accordion FAQ — was wrongly pruned as "framework content" but contains no React/CLI specifics',
    installSection: false,
    pinned: true, // top of sidebar, above the components list
  },
  {
    slug: 'installation',
    route: '/docs/installation',
    title: 'Installation',
    source: 'docs/content/installation.mdx',
    disposition: 'adapted',
    notes: 'authored vanilla rewrite; replaces upstream installation/index.mdx (CLI/create/framework cards — React toolchain) and installation/manual.mdx (Tailwind+aliases+components.json setup); copy-files story over dist/ artifacts',
    installSection: false,
  },
  {
    slug: 'dark-mode',
    route: '/docs/dark-mode',
    title: 'Dark Mode',
    source: 'docs/content/dark-mode.mdx',
    disposition: 'adapted',
    notes: 'authored rewrite; upstream dark-mode/index.mdx is framework cards only (next/vite/astro/remix/tanstack-start pruned); .dark theme variables ship precompiled in dist/out.css; mode-toggle preview to-author (FT7)',
    installSection: false,
  },
  {
    slug: 'rtl',
    route: '/docs/rtl',
    title: 'RTL',
    source: `${UP}/rtl/index.mdx`,
    disposition: 'mirror',
    notes: 'load-bearing (56 radix pages link /docs/rtl); framework sub-pages (rtl/next|vite|start) pruned → their LinkedCard links greyed; card-rtl preview to-author (FT7); shadcn-CLI migrate section rewritten (rtlMigrate)',
    installSection: false,
    rtlMigrate: true,
  },
  {
    slug: 'shimmer',
    route: '/docs/utils/shimmer',
    title: 'shimmer',
    source: `${UP}/utils/shimmer.mdx`,
    disposition: 'adapted',
    notes: 'mirrored; Installation section replaced (utilities ship precompiled in dist/out.css, no npm install); all 9 previews base-style (base-rhea) → unavailable',
    installSection: true,
    util: 'shimmer',
  },
  {
    slug: 'scroll-fade',
    route: '/docs/utils/scroll-fade',
    title: 'scroll-fade',
    source: `${UP}/utils/scroll-fade.mdx`,
    disposition: 'adapted',
    notes: 'mirrored; Installation section replaced (utilities ship precompiled in dist/out.css); all 7 previews base-style (6 base-rhea + 1 base-nova) → unavailable',
    installSection: true,
    util: 'scroll-fade',
  },
  {
    slug: 'ai-sdk',
    route: '/docs/helpers/ai-sdk',
    title: 'AI SDK',
    source: `${UP}/helpers/ai-sdk.mdx`,
    disposition: 'mirror',
    notes: 'kept per keep-list; @shadcn/helpers/ai-sdk is a React useChat package — mirrored as reference (fences stay verbatim, same policy as radix pages); ai-sdk-helper-demo base-style → unavailable',
    installSection: false,
  },
  {
    slug: 'tanstack-ai',
    route: '/docs/helpers/tanstack-ai',
    title: 'TanStack AI',
    source: `${UP}/helpers/tanstack-ai.mdx`,
    disposition: 'mirror',
    notes: 'kept per keep-list; @shadcn/helpers/tanstack-ai is a React package — mirrored as reference; tanstack-ai-helper-demo base-style → unavailable',
    installSection: false,
  },
  {
    slug: 'typography',
    route: '/docs/typography',
    title: 'Typography',
    source: 'docs/content/typography.mdx',
    disposition: 'adapted',
    notes: 'FT8: vanilla rewrite — upstream typography.mdx demos a <Typography> component shadless does not (and should not) ship; this guide maps the same typographic roles to plain Tailwind utilities already in dist/out.css',
    installSection: false,
  },
]

export const PRUNED = {
  forms: { source: `${UP}/forms/`, reason: 'React-only (react-hook-form / tanstack-form / formisch guides); field.mdx links ×4 → greyed spans' },
  react: { source: `${UP}/react/`, reason: 'React-only component recipes (message-scroller, questionnaire); radix links ×6 → greyed spans' },
  registry: { source: `${UP}/registry/`, reason: 'shadcn CLI registry system (json schema, MCP, namespaces) — no vanilla equivalent' },
  changelog: { source: `${UP}/changelog/`, reason: 'shadcn release notes — not shadless content' },
  '(root)': { source: `${UP}/(root)/`, reason: 'shadcn-site root pages (theming, cli, components.json…) — React/CLI specific' },
  'framework sub-pages': { source: `${UP}/{installation,dark-mode,rtl}/* (non-index)`, reason: 'per-React-framework setup guides (next/vite/astro/remix/tanstack/laravel/gatsby…) — installation/dark-mode/rtl index pages kept instead' },
}

// ---- shared /docs/ route resolver ------------------------------------------------
// members: Set of component page slugs. Returns {file, frag} for routable
// targets, {grey:true} for pruned/unknown. Upstream variant routes
// (/docs/components/{base,aria}/<name>) fall through to grey: the
// base/aria mirror is retired (2026-08-26) and those pages don't exist.
export function resolveDocsRoute(href, members) {
  if (!href.startsWith('/') || href.startsWith('//')) return null
  const [path, frag] = href.split('#')
  const comp = /^\/docs\/components\/(?:radix\/)?([a-z0-9-]+)$/.exec(path)
  if (comp) return members.has(comp[1]) ? { file: `${comp[1]}.html`, frag } : { grey: true }
  const guide = GUIDES.find((g) => g.route === path)
  if (guide) return { file: `${guide.slug}.html`, frag }
  return { grey: true } // pruned guides, /create, anything else
}

// ---- guide ComponentPreview enumeration (same scan discipline as the catalog) ----
const stripFences = (text) => text.replace(/```[\s\S]*?```/g, (m) => m.replace(/[^\n]/g, ' '))

export function scanGuidePreviews(catalog) {
  const catalogPreviews = new Map(catalog.previews.map((p) => [p.name, p.status]))
  const out = {}
  for (const g of GUIDES) {
    const text = stripFences(readFileSync(g.source, 'utf8'))
    for (const m of text.matchAll(/<ComponentPreview\b([^>]*)>/g)) {
      const attrs = m[1]
      // (^|\s) anchor: a bare \b would let data-name= match name=
      const attr = (name) => (attrs.match(new RegExp(`(?:^|\\s)${name}="([^"]*)"`)) || [])[1]
      const name = attr('name')
      if (!name) continue
      if (out[name]) { out[name].hostPages.push(g.slug); continue }
      const styleName = attr('styleName') || null
      let disposition, reason
      if (catalogPreviews.has(name)) {
        disposition = catalogPreviews.get(name) // e.g. card-rtl → to-author (FT7)
        reason = 'already cataloged in the radix set'
      } else if (styleName && styleName.startsWith('base-')) {
        disposition = 'unavailable'
        reason = `base-line demo (${styleName}) — shadless implements the radix line only`
      } else {
        disposition = 'to-author'
        reason = 'authored in an FT7 wave'
      }
      out[name] = { hostPages: [g.slug], styleName, disposition, reason }
    }
  }
  return out
}

// ---- content-map ------------------------------------------------------------------
// docs/content-map.json covers EVERY mirrored page (49 components + 9 guides
// + generated index) with a disposition, plus the pruned record and the guide
// preview table.
export function writeContentMap(componentPages) {
  const catalog = JSON.parse(readFileSync('docs/catalog.json', 'utf8'))
  const pages = {}
  for (const c of componentPages) {
    pages[c.name] = {
      source: c.source,
      disposition: 'adapted',
      notes: 'radix mirror; installation Manual tab rewritten to the vanilla copy-files path (build-time transform over dist/ artifacts)',
    }
  }
  for (const g of GUIDES) pages[g.slug] = { source: g.source, disposition: g.disposition, notes: g.notes }
  pages.index = { source: '(generated)', disposition: 'generated', notes: 'components + guides index page' }

  const contentMap = {
    version: 1,
    generatedBy: 'tools/docs-guides.mjs (FT4)',
    pages,
    pruned: PRUNED,
    guidePreviews: scanGuidePreviews(catalog),
  }
  mkdirSync('docs', { recursive: true })
  writeFileSync('docs/content-map.json', JSON.stringify(contentMap, null, 2) + '\n')
  return contentMap
}

// ---- CLI ---------------------------------------------------------------------------
if (process.argv[1] && basename(process.argv[1]) === 'docs-guides.mjs') {
  const catalog = JSON.parse(readFileSync('docs/catalog.json', 'utf8'))
  // Same radix-only page list the build uses (docs-build.mjs) — the CLI
  // previously rebuilt it radix-only and rewrote docs/content-map.json
  // with 98 fewer pages, silently degrading the docs-build cross-checks
  // until the next full build
  const componentPages = catalog.sources
    .filter((s) => s.status === 'existing-dist')
    .filter((s) => existsSync(join(`${UP}/components/radix`, `${s.name}.mdx`)))
    .map((s) => ({
      name: s.name,
      source: join(`${UP}/components/radix`, `${s.name}.mdx`),
    }))
    .sort((a, b) => a.name.localeCompare(b.name))
  const map = writeContentMap(componentPages)
  const by = (d) => Object.values(map.pages).filter((p) => p.disposition === d).length
  const guidesBy = (d) => GUIDES.filter((g) => g.disposition === d).length
  const nUnavail = Object.values(map.guidePreviews).filter((p) => p.disposition === 'unavailable').length
  const nToAuthor = Object.values(map.guidePreviews).filter((p) => p.disposition === 'to-author').length
  console.log(`guides keep-list: ${GUIDES.length} kept (${guidesBy('mirror')} mirror, ${guidesBy('adapted')} adapted), ${Object.keys(PRUNED).length} pruned groups`)
  console.log(`guide previews: ${Object.keys(map.guidePreviews).length} names (${nUnavail} unavailable base-style, ${nToAuthor} to-author)`)
  console.log(`content-map: docs/content-map.json (${Object.keys(map.pages).length} pages = ${componentPages.length} adapted component variants + ${guidesBy('adapted')} adapted guides + ${guidesBy('mirror')} mirrored guides + ${by('generated')} generated index)`)
}

File: pipeline/docs_guides.go
package main

// docs-guides, ported from tools/docs-guides.mjs — FT4 keep-list (decision
// record), the shared /docs/ route resolver, guide preview enumeration and
// the content-map writer. Merged here rather than into docs_catalog.go: the
// catalog is radix-only by design and lives on its own cadence; the guide
// table travels with the docs build that consumes it.

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

const guidesUp = ".upstream/shadcn-ui/apps/v4/content/docs"

type guide struct {
	slug           string
	route          string
	title          string
	source         string
	disposition    string
	notes          string
	installSection bool
	util           string
	rtlMigrate     bool
	pinned         bool
}

var guides = []guide{
	{slug: "introduction", route: "/docs/introduction", title: "Introduction", source: "docs/content/introduction.mdx", disposition: "adapted",
		notes:          "FT8: vanilla rewrite of upstream (root)/index.mdx; pure prose + 1 Accordion FAQ — was wrongly pruned as \"framework content\" but contains no React/CLI specifics",
		installSection: false, pinned: true},
	{slug: "installation", route: "/docs/installation", title: "Installation", source: "docs/content/installation.mdx", disposition: "adapted",
		notes:          "authored vanilla rewrite; replaces upstream installation/index.mdx (CLI/create/framework cards — React toolchain) and installation/manual.mdx (Tailwind+aliases+components.json setup); copy-files story over dist/ artifacts",
		installSection: false},
	{slug: "dark-mode", route: "/docs/dark-mode", title: "Dark Mode", source: "docs/content/dark-mode.mdx", disposition: "adapted",
		notes:          "authored rewrite; upstream dark-mode/index.mdx is framework cards only (next/vite/astro/remix/tanstack-start pruned); .dark theme variables ship precompiled in dist/out.css; mode-toggle preview to-author (FT7)",
		installSection: false},
	{slug: "rtl", route: "/docs/rtl", title: "RTL", source: guidesUp + "/rtl/index.mdx", disposition: "mirror",
		notes:          "load-bearing (56 radix pages link /docs/rtl); framework sub-pages (rtl/next|vite|start) pruned → their LinkedCard links greyed; card-rtl preview to-author (FT7); shadcn-CLI migrate section rewritten (rtlMigrate)",
		installSection: false, rtlMigrate: true},
	{slug: "shimmer", route: "/docs/utils/shimmer", title: "shimmer", source: guidesUp + "/utils/shimmer.mdx", disposition: "adapted",
		notes:          "mirrored; Installation section replaced (utilities ship precompiled in dist/out.css, no npm install); all 9 previews base-style (base-rhea) → unavailable",
		installSection: true, util: "shimmer"},
	{slug: "scroll-fade", route: "/docs/utils/scroll-fade", title: "scroll-fade", source: guidesUp + "/utils/scroll-fade.mdx", disposition: "adapted",
		notes:          "mirrored; Installation section replaced (utilities ship precompiled in dist/out.css); all 7 previews base-style (6 base-rhea + 1 base-nova) → unavailable",
		installSection: true, util: "scroll-fade"},
	{slug: "ai-sdk", route: "/docs/helpers/ai-sdk", title: "AI SDK", source: guidesUp + "/helpers/ai-sdk.mdx", disposition: "mirror",
		notes:          "kept per keep-list; @shadcn/helpers/ai-sdk is a React useChat package — mirrored as reference (fences stay verbatim, same policy as radix pages); ai-sdk-helper-demo base-style → unavailable",
		installSection: false},
	{slug: "tanstack-ai", route: "/docs/helpers/tanstack-ai", title: "TanStack AI", source: guidesUp + "/helpers/tanstack-ai.mdx", disposition: "mirror",
		notes:          "kept per keep-list; @shadcn/helpers/tanstack-ai is a React package — mirrored as reference; tanstack-ai-helper-demo base-style → unavailable",
		installSection: false},
	{slug: "typography", route: "/docs/typography", title: "Typography", source: "docs/content/typography.mdx", disposition: "adapted",
		notes:          "FT8: vanilla rewrite — upstream typography.mdx demos a <Typography> component shadless does not (and should not) ship; this guide maps the same typographic roles to plain Tailwind utilities already in dist/out.css",
		installSection: false},
}

var prunedGuides = map[string]struct{ Source, Reason string }{
	"forms":                {guidesUp + "/forms/", "React-only (react-hook-form / tanstack-form / formisch guides); field.mdx links ×4 → greyed spans"},
	"react":                {guidesUp + "/react/", "React-only component recipes (message-scroller, questionnaire); radix links ×6 → greyed spans"},
	"registry":             {guidesUp + "/registry/", "shadcn CLI registry system (json schema, MCP, namespaces) — no vanilla equivalent"},
	"changelog":            {guidesUp + "/changelog/", "shadcn release notes — not shadless content"},
	"(root)":               {guidesUp + "/(root)/", "shadcn-site root pages (theming, cli, components.json…) — React/CLI specific"},
	"framework sub-pages":  {guidesUp + "/{installation,dark-mode,rtl}/* (non-index)", "per-React-framework setup guides (next/vite/astro/remix/tanstack/laravel/gatsby…) — installation/dark-mode/rtl index pages kept instead"},
}

// routeTarget is resolveDocsRoute's result: routable (file+frag) or grey.
type routeTarget struct {
	file string
	frag string
	grey bool
}

var reCompRoute = regexp.MustCompile(`^/docs/components/(?:radix/)?([a-z0-9-]+)$`)

func resolveDocsRoute(href string, members map[string]bool) *routeTarget {
	if !strings.HasPrefix(href, "/") || strings.HasPrefix(href, "//") {
		return nil
	}
	path := href
	frag := ""
	if i := strings.IndexByte(href, '#'); i >= 0 {
		path, frag = href[:i], href[i+1:]
	}
	if m := reCompRoute.FindStringSubmatch(path); m != nil {
		if members[m[1]] {
			return &routeTarget{file: m[1] + ".html", frag: frag}
		}
		return &routeTarget{grey: true}
	}
	for _, g := range guides {
		if g.route == path {
			return &routeTarget{file: g.slug + ".html", frag: frag}
		}
	}
	return &routeTarget{grey: true}
}

// scanGuidePreviews enumerates guide ComponentPreview names with the same
// fence-stripped tag-scoped discipline as the catalog.
var (
	reGuidePreviewTag = regexp.MustCompile(`<ComponentPreview\b([^>]*)>`)
	reGuideFences     = regexp.MustCompile("(?s)```.*?```")
)

func scanGuidePreviews(catalogPreviewStatus map[string]string) (map[string]guidePreviewInfo, []string) {
	out := map[string]guidePreviewInfo{}
	var order []string
	for _, g := range guides {
		b, err := os.ReadFile(g.source)
		if err != nil {
			continue
		}
		text := reGuideFences.ReplaceAllStringFunc(string(b), func(m string) string {
			return strings.Map(func(r rune) rune {
				if r == '\n' {
					return '\n'
				}
				return ' '
			}, m)
		})
		for _, m := range reGuidePreviewTag.FindAllStringSubmatch(text, -1) {
			name := attrOfJS(m[1], "name")
			if name == "" {
				continue
			}
			if e, seen := out[name]; seen {
				e.hostPages = append(e.hostPages, g.slug)
				out[name] = e
				continue
			}
			styleName := attrOfJS(m[1], "styleName")
			var disposition, reason string
			if st, ok := catalogPreviewStatus[name]; ok {
				disposition = st
				reason = "already cataloged in the radix set"
			} else if strings.HasPrefix(styleName, "base-") {
				disposition = "unavailable"
				reason = "base-line demo (" + styleName + ") — shadless implements the radix line only"
			} else {
				disposition = "to-author"
				reason = "authored in an FT7 wave"
			}
			out[name] = guidePreviewInfo{hostPages: []string{g.slug}, styleName: styleName, disposition: disposition, reason: reason}
			order = append(order, name)
		}
	}
	return out, order
}

type guidePreviewInfo struct {
	hostPages   []string
	styleName   string
	disposition string
	reason      string
}

// attrOfJS: (^|\s)name="value" — a bare \b would let data-name= match name=.
func attrOfJS(attrs, name string) string {
	re := regexp.MustCompile(`(?:^|\s)` + regexp.QuoteMeta(name) + `="([^"]*)"`)
	if m := re.FindStringSubmatch(attrs); m != nil {
		return m[1]
	}
	return ""
}

// writeContentMap emits docs/content-map.json byte-compatibly with the JS
// writer (JSON.stringify(_, null, 2) + "\n"; key order = insertion order).
func writeContentMap(componentPages []struct{ name, source string }, sections map[string][]string) error {
	catalogB, err := os.ReadFile("docs/catalog.json")
	if err != nil {
		return err
	}
	var catalog struct {
		Previews []struct {
			Name   string `json:"name"`
			Status string `json:"status"`
		} `json:"previews"`
	}
	if err := json.Unmarshal(catalogB, &catalog); err != nil {
		return err
	}
	status := map[string]string{}
	for _, p := range catalog.Previews {
		status[p.Name] = p.Status
	}
	gp, gpOrder := scanGuidePreviews(status)

	// pages: components (sorted), then guides, then index — the JS writer's
	// insertion order. sections land LAST inside each entry (the JS build
	// appended them in a second pass).
	var pagesKV []jsonKV
	for _, c := range componentPages {
		e := []jsonKV{
			{"source", c.source},
			{"disposition", "adapted"},
			{"notes", "radix mirror; installation Manual tab rewritten to the vanilla copy-files path (build-time transform over dist/ artifacts)"},
		}
		if sec := sections[c.name]; len(sec) > 0 {
			arr := make([]any, len(sec))
			for i, x := range sec {
				arr[i] = x
			}
			e = append(e, jsonKV{"sections", arr})
		}
		pagesKV = append(pagesKV, jsonKV{c.name, e})
	}
	for _, g := range guides {
		e := []jsonKV{
			{"source", g.source},
			{"disposition", g.disposition},
			{"notes", g.notes},
		}
		if sec := sections[g.slug]; len(sec) > 0 {
			arr := make([]any, len(sec))
			for i, x := range sec {
				arr[i] = x
			}
			e = append(e, jsonKV{"sections", arr})
		}
		pagesKV = append(pagesKV, jsonKV{g.slug, e})
	}
	pagesKV = append(pagesKV, jsonKV{"index", []jsonKV{
		{"source", "(generated)"},
		{"disposition", "generated"},
		{"notes", "components + guides index page"},
	}})

	// pruned in the recorded order
	pkeys := []string{"forms", "react", "registry", "changelog", "(root)", "framework sub-pages"}
	var prunedKV []jsonKV
	for _, k := range pkeys {
		p := prunedGuides[k]
		prunedKV = append(prunedKV, jsonKV{k, []jsonKV{
			{"source", p.Source},
			{"reason", p.Reason},
		}})
	}

	// guidePreviews in first-seen order
	var gpKV []jsonKV
	for _, name := range gpOrder {
		g := gp[name]
		var style any
		if g.styleName != "" {
			style = g.styleName
		}
		hosts := make([]any, len(g.hostPages))
		for i, h := range g.hostPages {
			hosts[i] = h
		}
		gpKV = append(gpKV, jsonKV{name, []jsonKV{
			{"hostPages", hosts},
			{"styleName", style},
			{"disposition", g.disposition},
			{"reason", g.reason},
		}})
	}

	root := []jsonKV{
		{"version", 1},
		{"generatedBy", "tools/docs-guides.mjs (FT4)"},
		{"pages", pagesKV},
		{"pruned", prunedKV},
		{"guidePreviews", gpKV},
	}
	if err := os.MkdirAll("docs", 0o755); err != nil {
		return err
	}
	return os.WriteFile("docs/content-map.json", []byte(stringifyJSON(root, 0)+"\n"), 0o644)
}

// runDocsGuides is the CLI entry (the old tools/docs-guides.mjs main).
func runDocsGuides() int {
	catalogB, err := os.ReadFile("docs/catalog.json")
	if err != nil {
		fmt.Fprintln(os.Stderr, "docs-guides:", err)
		return 1
	}
	var catalog struct {
		Sources []struct {
			Name   string `json:"name"`
			Status string `json:"status"`
		} `json:"sources"`
	}
	if err := json.Unmarshal(catalogB, &catalog); err != nil {
		fmt.Fprintln(os.Stderr, "docs-guides:", err)
		return 1
	}
	var componentPages []struct{ name, source string }
	for _, s := range catalog.Sources {
		if s.Status != "existing-dist" {
			continue
		}
		p := filepath.Join(guidesUp+"/components/radix", s.Name+".mdx")
		if _, err := os.Stat(p); err != nil {
			continue
		}
		componentPages = append(componentPages, struct{ name, source string }{s.Name, p})
	}
	sort.Slice(componentPages, func(i, j int) bool {
		return componentPages[i].name < componentPages[j].name
	})
	if err := writeContentMap(componentPages, nil); err != nil {
		fmt.Fprintln(os.Stderr, "docs-guides:", err)
		return 1
	}
	nMirror, nAdapted := 0, 0
	for _, g := range guides {
		if g.disposition == "mirror" {
			nMirror++
		} else {
			nAdapted++
		}
	}
	fmt.Printf("guides keep-list: %d kept (%d mirror, %d adapted), %d pruned groups\n",
		len(guides), nMirror, nAdapted, len(prunedGuides))
	return 0
}
